hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ConditionalWriteLockService.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.client.transaction.lock;
+
+import org.apache.hudi.client.transaction.lock.models.ConditionalWriteLockData;
+import org.apache.hudi.client.transaction.lock.models.ConditionalWriteLockFile;
+import org.apache.hudi.client.transaction.lock.models.LockGetResult;
+import org.apache.hudi.client.transaction.lock.models.LockUpdateResult;
+import org.apache.hudi.common.util.collection.Pair;
+
+import java.util.function.Supplier;
+
+/**
+ * Defines a contract for a service which should be able to perform conditional writes to object storage.
+ * It expects to be interacting with a single lock file per context (table), and will be competing with other instances
+ * to perform writes, so it should handle these cases accordingly (using conditional writes).
+ */
+public interface ConditionalWriteLockService extends AutoCloseable {
```

Comment:
naming: just `ConditionalWriteLock` ? Service implies sth is started/closed and there is a lifecycle. 

also should this be called just "StorageLock" - thats the main difference compared to say ZK. that this is implemented right on top of Hudi's backing storage? "Writer" indicates that this is how we update/write to the underlying lock

Reply:
I added "Service" because the implementations are sort of service-y but you're right there is no lifecycle.

Reply:
I think I like `StorageLock` better than `ConditionalWriteLock`